src/transport/tests/TestSecureSessionMgr.cpp

```diff
@@ -400,6 +402,64 @@ void SendBadEncryptedPacketTest(nlTestSuite * inSuite, void * inContext)
     NL_TEST_ASSERT(inSuite, callback.ReceiveHandlerCallCount == 2);
 }
 
+void ConnectionExpiryTest(nlTestSuite * inSuite, void * inContext)
+{
+    TestContext & ctx = *reinterpret_cast<TestContext *>(inContext);
+
+    IPAddress addr;
+    IPAddress::FromString("127.0.0.1", addr);
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    TransportMgr<LoopbackTransport> transportMgr;
+    SecureSessionMgr secureSessionMgr;
+    secure_channel::MessageCounterManager gMessageCounterManager;
+
+    err = transportMgr.Init("LOOPBACK");
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+
+    Transport::FabricTable fabrics;
+    err = secureSessionMgr.Init(ctx.GetInetLayer().SystemLayer(), &transportMgr, &fabrics, &gMessageCounterManager);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+
+    callback.mSuite = inSuite;
+
+    secureSessionMgr.SetDelegate(&callback);
+
+    Optional<Transport::PeerAddress> peer(Transport::PeerAddress::UDP(addr, CHIP_PORT));
+
+    // First pairing
+    SecurePairingUsingTestSecret pairing1(1, 1);
+    err = secureSessionMgr.NewPairing(peer, kSourceNodeId, &pairing1, SecureSession::SessionRole::kInitiator, 1);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, !callback.mConnectionExpired);
+
+    // New pairing with different peer node ID and different local key ID (same peer key ID)
+    SecurePairingUsingTestSecret pairing2(1, 2);
+    err = secureSessionMgr.NewPairing(peer, kSourceNodeId, &pairing2, SecureSession::SessionRole::kResponder, 0);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, !callback.mConnectionExpired);
+
+    // New pairing with undefined node ID and different local key ID (same peer key ID)
+    SecurePairingUsingTestSecret pairing3(1, 3);
+    err = secureSessionMgr.NewPairing(peer, kUndefinedNodeId, &pairing3, SecureSession::SessionRole::kResponder, 0);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, !callback.mConnectionExpired);
+
+    // New pairing with same local key ID, and a given node ID
+    SecurePairingUsingTestSecret pairing4(1, 2);
+    err = secureSessionMgr.NewPairing(peer, kSourceNodeId, &pairing4, SecureSession::SessionRole::kResponder, 0);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, callback.mConnectionExpired);
+
+    callback.mConnectionExpired = false;
```

Comment:
Since tests don't abort on assertion failure, suggest setting this to false before each-subtest-case so that if there is one failure, the statefulness above is not assumed to mean success.